### PR TITLE
Cloud backup instrumentation

### PR DESCRIPTION
**Pull request by digitarald**
Shortlist of well known cloud drive providers

### CODE REVIEW

**Review by lramos15 — DISMISSED** [review]
👍

File: src/vs/platform/diagnostics/node/diagnosticsService.ts [hunk 1]
```diff
@@ -61,12 +66,27 @@ export async function collectWorkspaceStats(folder: string, filter: string[]): P
 		{ tag: 'dockerfile', filePattern: /^(dockerfile|docker\-compose\.ya?ml)$/i }
 	];
 
+	const rootPathPatterns: RootFilePatterns[] = [
```
Comment (ItalyPaleAle) [inline src/vs/platform/diagnostics/node/diagnosticsService.ts:69]: Uhm I am worried these regex may create a bunch of false positives, especially more generically-named ones like `box`: they would be triggered for any folder called `box`. Instead, it should be `~/Box` on macOS and `%HOMEPATH%/Box` on Windows.

**Review by ItalyPaleAle — DISMISSED** [review]

**Comment by ItalyPaleAle** [comment]
I changed the detection logic so it's stricter and matches the paths we found in our research.

I also made it OS-specific to simplify the logic (esp. given the different path separators and some other OS-specific differences). Because we don't have data on where things are stored on Linux, nothing should be collected there - there's also the fact that on Linux there are multiple ways to add cloud drives (including some built into the shells, like with GNOME).

Lastly, please note that for some reason I can't get VS Code to build on my laptop so I couldn't test this on macOS. It works on Codespaces but that's Linux-based so it won't collect anything

File: src/vs/platform/diagnostics/node/diagnosticsService.ts [hunk 2]
```diff
@@ -61,12 +66,146 @@ export async function collectWorkspaceStats(folder: string, filter: string[]): P
...
+			];
+			break;
+
+		case 'darwin':
+			rootFileMatchers = [
+				{
+					tag: 'gdrive', matcher: (path) => {
+						path = path.toLowerCase();
+						// File Streaming mode
+						if (path.startsWith('/volumes/googledrive/')) {
```
Comment (lramos15) [inline src/vs/platform/diagnostics/node/diagnosticsService.ts:134]: can't these be simplified to `return path.startsWith('/volumes/googledrive/') || path.startsWith(homeDir + '/my drive/');`

**Review by lramos15 — APPROVED** [review]
See my suggestion about condensing some of the ifs, but that may be less clear in terms of readability so up to you.

**Comment by JacksonKearl** [comment]
Verified on macOS the event is correctly sent when opening a folder in iCloud drive, verified the rest via inspection of the telemetry:
image

I'm somewhat surprised OneDrive is so much more popular that gDrive, and wonder if that may be an issue in how we identify gDrive instances. Or OneDrive may really be that popular, I'm not sure.

**Comment by ItalyPaleAle** [comment]
Uhmmm - I wonder what happens if someone on Windows is storing their Desktop or Home folder on OneDrive. Would that appear here too?

**Comment by JacksonKearl** [comment]
Not sure re OneDrive, though digging into the gdrive more I found that while we do capture the default location for mirrored mode, it is possible to choose any folder to be the mirror so we cannot capture those.

**Comment by ItalyPaleAle** [comment]
Let me run some tests with syncing the Home folder to OneDrive and see if I learn something.

As for non-default folders, yes: (almost) all the sync tools above allow selecting any folder. I believe the only exception is iCloud. I am not sure if investing in detecting non-standard folders is worth it at this stage, as it doesn't look like there's a convenient way to do that. (As per my research, for example, on Windows you can look up the folder Dropbox uses but it requires reading the Registry: https://superuser.com/questions/1473136/real-location-of-onedrive-folder-in-windows-10 )

**Comment by ItalyPaleAle** [comment]
@JacksonKearl It's what I was suspecting. When someone is syncing their Desktop, Documents, or Photos folders on Windows 10/11, they are actually operating inside the OneDrive folder.

One thing we could try is including `%HOMEPATH%/OneDrive*` but excluding `%HOMEPATH%/OneDrive*/Documents` (and Desktop and Photos).

The only question I have in that case is what happens if the user is using Windows in a different language than English.

**Comment by ItalyPaleAle** [comment]
I also wonder if the same could happen with iCloud on Mac, since that too seems suspiciously high

**Comment by ItalyPaleAle** [comment]
Just chatted with @JacksonKearl and we think this is a desired behavior, as we want to know if people are using cloud backups regardless.

When querying the data, we will then exclude any workspace that is using SCM.

**Pull request merged**